Userland/Libraries/LibWeb/HTML/Parser/HTMLToken.h

```diff
@@ -291,25 +386,29 @@ class HTMLToken {
     void set_end_position(Badge<HTMLTokenizer>, Position end_position) { m_end_position = move(end_position); }
 
 private:
+    Vector<Attribute>& ensure_tag_attributes()
+    {
+        VERIFY(is_start_tag() || is_end_tag());
+        if (!m_tag_attributes)
+            m_tag_attributes = new Vector<Attribute>;
+        return *m_tag_attributes;
+    }
+
     Type m_type { Type::Invalid };
 
-    // Type::DOCTYPE
-    DoctypeData m_doctype;
-
-    // Type::StartTag
-    // Type::EndTag
-    struct {
-        String tag_name;
-        bool self_closing { false };
-        bool self_closing_acknowledged { false };
-        Vector<Attribute> attributes;
-    } m_tag;
-
-    // Type::Comment
-    // Type::Character
-    struct {
-        String data;
-    } m_comment_or_character;
+    // Type::Comment (comment data), Type::StartTag and Type::EndTag (tag name)
+    String m_string_data;
+
+    // Type::StarTag and Type::EndTag
+    bool m_tag_self_closing { false };
+    bool m_tag_self_closing_acknowledged { false };
+
+    // NOTE: We need to initialize one of the pointers by default, since they can be larger than u32, but never smaller.
+    union {
+        u32 m_code_point;                        // Type::Character
+        DoctypeData* m_doctype_data { nullptr }; // Type::DOCTYPE
+        Vector<Attribute>* m_tag_attributes;     // Type::StartTag and Type::EndTag
+    };
```

Comment:
I'd also suggest moving some of the `bool`s to the end, since they're probably adding useless padding.
Do experiment with it though, it might be okay as-is.